[test] Basic Auth Identity Provider: Handle redirection safely

sgallagher opened the pull request:
@liggitt 
Related to https://github.com/openshift/origin/issues/4977

These patches add several new tests to the integration suite.

The first patch ensures that incorrect passwords are properly rejected and return a 401 from `RequestToken()`.

The second patch adds tests for when a remote Basic Auth server returns a redirect code 301, 302, 303 or 307.

The third patch fixes a bug in the Basic Auth provider where it was automatically following redirects. We don't want to do this because this may have been an attempt to redirect authentication to a malicious site. It's better to simply reject and let an administrator figure out what the correct behavior would be.

liggitt commented on pkg/auth/authenticator/password/basicauthpassword/basicauthpassword.go:
```diff
@@ -60,6 +67,13 @@ func New(providerName string, url string, transport http.RoundTripper, mapper au
 	}
 	client := &http.Client{Transport: transport}
+
+	// We don't support redurects in the basic auth provider because it could be a malicious attempt to send credentials to
```
typo

liggitt commented on pkg/auth/authenticator/password/basicauthpassword/basicauthpassword.go:
```diff
@@ -52,6 +52,13 @@ type RemoteError struct {
 }
 
+// Error type indicating the type of redirect that was detected
+type RedirectAttemptedError string
```
can just make `var RedirectAttempted = errors.New("redirect attempted")` and return it directly

liggitt commented on pkg/auth/authenticator/password/basicauthpassword/basicauthpassword.go:
```diff
@@ -80,6 +94,11 @@ func (a *Authenticator) AuthenticatePassword(username, password string) (user.In
 	if err != nil {
+		if _, ok := err.(*RedirectAttemptedError); !ok {
+			// If this was a redirect, treat it as a simple
+			// authentication failure, not an error
```
why wouldn't this be an error? the remote site is behaving outside spec, right?

sgallagher replied:
@liggitt I guess it comes down to how we handle errors. Right now, if this is returning an actual error, then what the end-user gets back is a 500 error from the `tokenRequest()` call which replaces this error with a different one about refusing to retry. That's not a good user experience for a number of reasons.

To me, a 500 error means that there is a bug in the code somewhere. A 500 error coming out of a *handled* error condition is incorrect. If we want to report a proper error code, we need to decide which one is most appropriate here. At a glance, I'd say it should be one of 404, 410 or 421 (or possibly 418 to be a little tongue-in-cheek).

Thoughts?

liggitt replied:
If the idp is not returning a 401 or 200, it is misconfigured, and I think the user should get a 500.

liggitt replied:
Does the 500 make it all the way to the user, or does the oauth layer change it to a different status code?

sgallagher replied:
The 500 error is returned by `tokencmd.RequestToken()` to the user. I'm not sure how exactly it behaves in the web UI yet.

So how do you want me to proceed here?

liggitt commented on test/integration/oauth_basicauth_test.go:
```diff
@@ -288,7 +309,151 @@ func TestOAuthBasicAuthPassword(t *testing.T) {
+	runRedirectTests(t, 303)
+}
+func TestOAuthBasicAuthPasswordRedirect307(t *testing.T) {
+	runRedirectTests(t, 307)
```
the tests take ~8-10 seconds apiece... I would just do a single redirect test

sgallagher replied:
Actually, I trimmed this down, but I noticed that we don't gracefully handle other status codes in the 3xx range. (Originally, this listed 300-308, but I put the others aside for this PR). The `RequestToken()` call returns a 500 error in those cases. I was going to discuss with you whether that's correct or if we should handle that better.

sgallagher replied:
So are you suggesting I just pick one of the codes and test that, or modify the test so I only start and stop a single instance of OpenShift?

liggitt replied:
I would modify the test to let you tweak the response from the proxy in a loop that hits all the return code types we care about

sgallagher commented:
@liggitt PTAL

liggitt commented on test/integration/oauth_basicauth_test.go:
```diff
@@ -32,7 +33,12 @@ const (
 
 var (
-	basicAuthCerts = map[string][]byte{
+	remotePrefix       = "remote"
```
is there a way to avoid lots of package-level vars? these get shared across all integration tests

sgallagher replied:
I don't see a good way to do it without duplication which would be hard to keep in sync. I'm removing remotePrefix though, since it was only being used in another package var anyway.

liggitt commented on test/integration/oauth_basicauth_test.go:
```diff
@@ -195,35 +223,17 @@ func TestOAuthBasicAuthPassword(t *testing.T) {
 		if req.Header.Get("Authorization") != expectedAuthHeader {
 			w.WriteHeader(http.StatusUnauthorized)
-			t.Fatalf("Unexpected auth header: %s", req.Header.Get("Authorization"))
+			return
```
why aren't we failing the test here now?

sgallagher replied:
This is part of the first of the three patches. I added functionality to do a negative test: proof that attempting to log in with the wrong password would be blocked. In order to do that, we needed to have the auth server return `http.StatusUnauthorized` rather than failing the test.

See below.

liggitt commented on test/integration/oauth_basicauth_test.go:
```diff
@@ -288,7 +309,149 @@ func TestOAuthBasicAuthPassword(t *testing.T) {
+}
+
+var redirectCodes = []int{300, 301, 302, 303, 304, 305, 306, 307}
+var redirectCodeIndex = 0
```
I'd rather not move to static package-level vars to drive this behavior, updating the remote basic auth handler closure should be sufficient

sgallagher commented:
@liggitt I've rebased atop the latest master and moved the commits around; now there's one commit to add the redirection prevention in the auth code and a separate commit that adds all of the new tests, merging your patch from https://github.com/liggitt/origin/commit/c519a0b3df2f401ca6bb432ed9045a5835946695 and my original negative auth test, fixing up a couple minor bugs in your patch at the same time.

liggitt commented:
LGTM, go ahead and squash

sgallagher commented:
I don't see any reason to squash these two patches together. They're both internally consistent and logically separate. Leaving them independent would seem preferable for the git history.

sgallagher commented:
@liggit I squashed them as requested.

liggitt commented:
[merge]
